### PR TITLE
Replace arg_policy.team_size_max with Kokkos::Impl::cuda_get_max_block_size in Cuda::ParallelReduce

### PR DESCRIPTION
**Pull request by janciesko**
Fixes issue #4095.

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 1]
```diff
@@ -411,10 +411,12 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
...
-    int p2 = 1;
-    while (p2 <= block_size) p2 *= 2;
-    p2 /= 2;
-    return p2 / impl_vector_length();
+    int pow2 = 0;
+    int block_size_pow2 = block_size;
+    while (block_size_pow2 >>= 1 > 0 ) pow2++;
+    block_size_pow2 = 1 << pow2;
+
+    return block_size_pow2 / impl_vector_length();
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:419]: Does this make any difference? AFAICT (https://godbolt.org/z/x6v1MEWb6) they show the same behavior.

**Comment by DavidPoliakoff** [comment]
Yeah, I'm also missing it. Give us a comment saying what the bug is, and how this fixes it?

**Comment by masterleinad** [comment]
> Yeah, I'm also missing it. Give us a comment saying what the bug is, and how this fixes it?

I'm pretty sure that https://github.com/kokkos/kokkos/pull/4142/files#diff-99024452757b67f0dea936c050017c64f4ec289c2fac1a463689b39e653891edR2007-R2012 alone fixes the issue. The question for me is if we want to silently fix the team size to be a power of two or error out (and only fix it if `AUTO` is given).

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 1 above)
Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:419]: Change is for consistency following the added code block to compute pow2 in the ParallelReduce constructor. In case of using Kokkos::Auto in the ExecutionPolicy for that parallel_reduce, cuda_get_opt_block_size returns a non-pow2 team size. Setting the team size to the next smallest pow2, line 2007, fixes the reported issue.

**Comment by janciesko** [comment]
We could do that. If (m_team_size==-1), compute the team size (cuda_get_opt_block_size(...)) && set it to the smallest power of 2 value. I'm curious whether this requirement for this specific number of participating threads in a reduction exists for all Cuda archs since the issue has been reported for Kepler 3.7 only.

**Comment by janciesko** [comment]
Todo: Other ParallelReduce overloads might be affected as well.

**Comment by masterleinad** [comment]
Summarizing from the last team meeting, we decided to make sure that the team size calculated for `AUTO` is a power of two but not to change the team size otherwise.

**Comment by janciesko** [comment]
Retest this please.

**Review by masterleinad — APPROVED** [review]
The fix looks fine to me. I slightly prefer using the old style for the power-2 calculation to be used in the new check, though, just to change less code. I don't see that either implementation is better than the other.

**Review by Rombur — APPROVED** [review]

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 2]
```diff
@@ -410,11 +410,13 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
...
-    while (p2 <= block_size) p2 *= 2;
-    p2 /= 2;
-    return p2 / impl_vector_length();
+    // Currently we require power-of-two team size for reductions.
+    int pow2            = 0;
+    int block_size_pow2 = block_size;
+    while (block_size_pow2 >>= 1 > 0) pow2++;
+    block_size_pow2 = 1 << pow2;
+
+    return block_size_pow2 / impl_vector_length();
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:419]: why this change? The previous thing already produced a power of two didn't it?

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 3]
```diff
@@ -410,11 +410,13 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
                  : functor_value_traits::value_size(f)));
     KOKKOS_ASSERT(block_size > 0);
 
-    // Currently we require Power-of-2 team size for reductions.
-    int p2 = 1;
-    while (p2 <= block_size) p2 *= 2;
-    p2 /= 2;
-    return p2 / impl_vector_length();
+    // Currently we require power-of-two team size for reductions.
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:413]: I am not actually sure that this comment is still accurate.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 4]
```diff
@@ -1992,14 +1994,22 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
+              m_policy.space().impl_internal_space_instance(), attr, m_functor,
+              m_vector_size, m_policy.team_scratch_size(0),
+              m_policy.thread_scratch_size(0)) /
+          m_vector_size;
+
+      // Currently we require power-of-two team size for reductions.
+      int pow2 = 0;
+      while (m_team_size >>= 1 > 0) pow2++;
+      pow2        = 1 << pow2;
+      m_team_size = pow2;
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2011]: I don't think this is the right approach. I think the cuda_get_opt_block_size function needs an option to only return power of two (if that actually is the issue). Because you might defeat the purposes of opt block size with what you are doing here.

**Review by crtrott — CHANGES_REQUESTED** [review]
I don't think this is all fully right, at a minimum part of the logic i don't understand why it was rewritten at all, and in the other case I think we need to correct the opt block size function not fix it up after the fact.

**Comment by crtrott** [comment]
Yeah I looked more at this. It just sidesteps the problem. It throws right now in core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2058
```c++
    if (int(m_team_size) >
        arg_policy.team_size_max(m_functor, m_reducer, ParallelReduceTag()))
```

but m_team_size comes from:
```c++
    m_team_size =
        m_team_size >= 0
            ? m_team_size
            : Kokkos::Impl::cuda_get_opt_block_size<FunctorType, LaunchBounds>(
                  m_policy.space().impl_internal_space_instance(), attr,
                  m_functor, m_vector_size, m_policy.team_scratch_size(0),
                  m_policy.thread_scratch_size(0)) /
                  m_vector_size;
```

so cuda_get_opt_block_size()/vector_size returns a value larger than policy.team_size_max()??
That is an inconsistency we need to fix at the root of the issue, not just put a band-aid on. So the question is why does cuda_get_opt_block_size()/vector_size return a value larger than policy.team_size_max() - and who of the two is wrong?

Note the comment about power of two team size is old, we actually test team_size 33 for all reductions.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 4 above)
Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: Is m_vector_size always a pow2 number?

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 2 above)
Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: Reverted.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 3 above)
Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: We might consider opening another issue on this as this PR fixes the reported error.

**Comment by janciesko** [comment]
Indeed this was side-stepping the problem. 

@ndellingwood: Do you maybe remember why this was necessary:  d5b3921bfa3e28305f3a8717251d42e9d8b2a21b
This reduces the thread block size to the next smaller power-of-two value in an inconsistent way. This issue shows only on architectures where the max value is sufficiently small and such that the a call to cuda_get_opt_block_size exceeds that power-of-two value.

The problematic code:
https://github.com/kokkos/kokkos/blob/9cf0d5bc02513aece0a79c12d717714cf9345700/core/src/Cuda/Kokkos_Cuda_Parallel.hpp#L413-L417

**Comment by ndellingwood** [comment]
@janciesko here was the reasoning for the change:
https://github.com/kokkos/kokkos/issues/2146#issuecomment-493238549
Recap from that comment: unit tests were broken when using `team_size_max` vs `team_size_recommended`; `team_size_max` resulted in an adjusted block-size to a power-of-2 which seemed necessary for the reduction to perform correctly at that time; `team_size_recommended` had no block-size readjustment, so non-power-of-2 cases failed due to apparent incompatibility with the reduction algorithm

**Comment by janciesko** [comment]
Yea, I see the problem.

**Comment by janciesko** [comment]
Thanks @ndellingwood for commenting. 

The issue reported caught a case where `CUDA::cudaFuncGetAttributes` returns `maxThreadsPerBlock=768`. This is specific to Kepler and the particular kernel functor that evaluates to that value. Our deduction algorithm returns an optimal block size of 544 and max block size of 768. That is correct and the algorithm accounts for `regsPerMultiprocessor`. After the deduction, the max block size is then reduced to 512 as in d5b3921. This results in the failing assertion of `optimal block size <= maximal block size`. This is not covered by our unit test as our unit tests report maxThreadsPerBlock of 1024 and the deduced opt/max values of 128/1024.

Briefly looking at https://github.com/kokkos/kokkos/blob/e4831441fb213bd29485327b02cf1b54149a58a7/core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp#L881-L885
I see that we require a pow2 block size for TeamThreadRange (blockDim.y). One solution is to also adjust the optimal block size to the next lower or equal pow2 number similarly as in d5b3921. This would correspond the to PR, but I think we should move that pow2 computation somewhere else and have a condition flag passed along.

Interestingly, the bouds checked specialization returns that CUDA function attribute value, which explains why the issue has been reported only when configuring with Kokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 4 above)
Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: Yes. I'll add that. That is the issue.

**Comment by janciesko** [comment]
Retest this please.

**Comment by janciesko** [comment]
This updated PR updates the assert condition to support non power-of-two team sizes in Cuda::ParallelReduce.

**Comment by masterleinad** [comment]
Can you please summarize why this is the right solution? It looks like `team_size_max` is not the true upper limit now?

**Comment by janciesko** [comment]
Yes, ```arg_policy.team_size_max``` used in the conditional is not the true upper limit for parallel reduce as it returns a team size of the next smaller pow2 value. Earlier we use Kokkos::Impl::cuda_get_opt_block_size to derive the recommended team size which can return a non-pow2 team size. The properly check for the true upper limit, we need to replace ```arg_policy.team_size_max``` with ```Kokkos::Impl::cuda_get_max_block_size```.

**Comment by masterleinad** [comment]
OK. I'm still slightly confused by https://github.com/kokkos/kokkos/blob/e4831441fb213bd29485327b02cf1b54149a58a7/core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp#L881-L885. Do we have a power-of-two requirement for parallel_reduce or not?

If we don't, I would rather revert https://github.com/kokkos/kokkos/commit/d5b3921bfa3e28305f3a8717251d42e9d8b2a21b. Otherwise, we should let `Kokkos::Impl::cuda_get_opt_block_size` return a power-of-two as well (which is hopefully less that the maximum (power-of-two) block size.

Can you clarify this?

**Comment by janciesko** [comment]
No, we cannot revert. That path is used for parallel scans that still require pow2. Once we have a better implementation of parallel scan we can remove the pow2 requirement. Parallel reduce does not have a pow2 requirement. If we use ```Kokkos::Impl::cuda_get_opt_block_size``` then we should use ```Kokkos::Impl::cuda_get_max_block_size``` in the if condition.

**Comment by masterleinad** [comment]
OK. I think this is sensible. Let me just try to summarize where we would hit the power-of-two block size restriction https://github.com/kokkos/kokkos/blob/23cfe6f7edcdf66fe67a665eb6238ea23488100a/core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp#L652-L654
This one is called in
```
$ git grep -n "cuda_intra_block_reduce_scan<"
core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2232:      cuda_intra_block_reduce_scan<true, FunctorType, WorkTag>(
core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2518:      cuda_intra_block_reduce_scan<true, FunctorType, WorkTag>(
core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:809:  cuda_intra_block_reduce_scan<false, FunctorType, ArgTag>(
core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:849:    cuda_intra_block_reduce_scan<DoScan, FunctorType, ArgTag>(
core/src/Cuda/Kokkos_Cuda_Team.hpp:266:    Impl::cuda_intra_block_reduce_scan<true, Impl::CudaJoinFunctor<Type>, void>(
```
The first two come from `ParallelScan`, the last one from `team_scan`. The other two are called by `cuda_single_inter_block_reduce_scan2` which is called by `cuda_single_inter_block_reduce_scan` for if
```
!DoScan && ValueTraits::StaticValueSize > 0
```
is not true (meaning we do array reductions or are using scans).
`cuda_single_inter_block_reduce_scan` in turn is used in
```
$ git grep -n "cuda_single_inter_block_reduce_scan<"
core/src/Cuda/Kokkos_Cuda_Parallel.hpp:999:        if (cuda_single_inter_block_reduce_scan<false, ReducerTypeFwd,
core/src/Cuda/Kokkos_Cuda_Parallel.hpp:1337:    if (cuda_single_inter_block_reduce_scan<false, ReducerTypeFwd, WorkTagFwd>(
core/src/Cuda/Kokkos_Cuda_Parallel.hpp:1718:      if (cuda_single_inter_block_reduce_scan<false, FunctorType, WorkTag>(
core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2170:    cuda_single_inter_block_reduce_scan<true, FunctorType, WorkTag>(
core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2454:    cuda_single_inter_block_reduce_scan<true, FunctorType, WorkTag>(
```
The first one is `ParallelReduce` for `RangePolicy`, the second one is `ParallelReduce` for `MDRangePolicy`, the third one is `ParallelReduce` for `TeamPolicy` and the last two ones are `ParallelScan` again.
`ParallelReduce` in `TeamPolicy` executes that code path never, though. Instead, the `ShflReduction` implementation is always used.

To summarize, with the current implementation, we never care about the team size being a power of two unless we call team_scan (and this issue is captured in https://github.com/kokkos/kokkos/issues/4146).

**Review by masterleinad — APPROVED** [review]

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 5]
```diff
@@ -2053,8 +2055,15 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
       Kokkos::Impl::throw_runtime_exception(
           std::string("Kokkos::Impl::ParallelReduce< Cuda > bad team size"));
     }
-    if (int(m_team_size) >
-        arg_policy.team_size_max(m_functor, m_reducer, ParallelReduceTag())) {
+
+    size_type m_team_size_max =
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2059]: Also consider defining the variable before L1997 if you are already using it there in the ternary operator.

**Review by dalg24 — COMMENTED** [review]
Why did you chose to change the check rather than the ternary in the initialization of `m_team_size`?

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: Dropped the ```m_```. In the initialization we set ```m_team_size``` according to the policy type or argument. In the ternary we update ```m_team_size``` if necessary. In the conditional we check if ```m_team_size``` is within team_size_max. Team_size_max was previously a pow2 value which triggered the reported bug.

**Comment by janciesko** [comment]
Retest this please.

**Comment by dalg24** [comment]
> Why did you chose to change the check rather than the ternary in the initialization of `m_team_size`?

Ping

**Review by crtrott — APPROVED** [review]
